be/src/runtime/batch_write/txn_state_cache.h

```diff
@@ -0,0 +1,273 @@
+    // If true, means publish timeout happens, and should notify subscribers. If false,
+    // means its from poll should continue to wait.
+    bool _committed_status_from_fe{false};
+    int32_t _num_subscriber{0};
+    int32_t _num_waiting_finished_state{0};
+    int32_t _num_poll_failure{0};
+    bool _stopped{false};
+};
+
+inline int32_t TxnStateHandler::num_subscriber() {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    return _num_subscriber;
+}
+
+inline int32_t TxnStateHandler::num_waiting_finished_state() {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    return _num_waiting_finished_state;
+}
+
+inline TxnState TxnStateHandler::txn_state() {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    return _txn_state;
+}
+
+inline bool TxnStateHandler::committed_status_from_fe() {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    return _committed_status_from_fe;
+}
+
+inline int32_t TxnStateHandler::num_poll_failure() {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    return _num_poll_failure;
+}
+
+inline std::ostream& operator<<(std::ostream& os, TxnStateHandler& holder) {
+    os << holder.debug_string();
+    return os;
+}
+
+// A subscriber which will wait for the finished txn state. It holds a reference
+// to the entry of DynamicCache so that the cache will not evict the txn state.
+// The subscriber can call wait_finished_state() to wait for the finished state.
+class TxnStateSubscriber {
+public:
+    TxnStateSubscriber(TxnStateDynamicCache* cache, TxnStateDynamicCacheEntry* entry, std::string name)
+            : _cache(cache), _entry(entry), _name(std::move(name)){};
+
+    ~TxnStateSubscriber() {
+        _entry->value().unsubscribe();
+        _cache->release(_entry);
+    }
+
+    const std::string& name() const { return _name; }
+    StatusOr<TxnState> wait_finished_state(int64_t timeout_us);
+    TxnState current_state();
+    TxnStateDynamicCacheEntry* entry() { return _entry; }
+
+private:
+    TxnStateDynamicCache* _cache;
+    TxnStateDynamicCacheEntry* _entry;
+    std::string _name;
+};
+using TxnStateSubscriberPtr = std::unique_ptr<TxnStateSubscriber>;
+
+struct TxnStatePollTask {
+    int64_t txn_id;
+    std::string db;
+    std::string tbl;
+    AuthInfo auth;
+};
+
+// Schedule and execute txn state poll tasks. The poller uses a single thread
+// to schedule tasks according to their execution time, and submit them to the
+// thread pool to run which will send rpc to FE to get txn state.
+class TxnStatePoller {
+public:
+    TxnStatePoller(TxnStateCache* txn_state_cache, ThreadPoolToken* poll_token)
+            : _txn_state_cache(txn_state_cache), _poll_token(poll_token) {}
+    Status init();
+    // submit a task which should be executed after the delay time
+    void submit(const TxnStatePollTask& task, int64_t delay_ms);
+    void stop();
+
+    // For testing
+    bool is_txn_pending(int64_t txn_id);
```

Comment:
```suggestion
    bool TEST_is_txn_pending(int64_t txn_id);
```

have those interface start with TEST_xxx, or condition compiling them with #ifdef BE_TEST #endif

Reply:
fixed